### PR TITLE
Add GW Detectors

### PR DESCRIPTION
**Pull request by mcoughlin**
This PR adds GW detectors.

### CODE REVIEW

File: skyportal/handlers/api/gwdetector.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
+from marshmallow.exceptions import ValidationError
+from baselayer.app.access import permissions, auth_or_token
+
+from ..base import BaseHandler
+from ...models import GWDetector
+
+
+class GWDetectorHandler(BaseHandler):
+    @permissions(['Manage allocations'])
```
Comment (guynir42) [inline skyportal/handlers/api/gwdetector.py:9]: Is this the right permissions?

File: skyportal/handlers/api/gwdetector.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+        with self.Session() as session:
+
+            schema = GWDetector.__schema__()
+            if 'lat' not in data or 'lon' not in data:
+                return self.error('Missing latitude or longitude')
+            elif not isinstance(data['lat'], (int, float)) or not isinstance(
+                data['lon'], (int, float)
+            ):
+                return self.error('Latitude and longitude must all be numbers')
```
Comment (guynir42) [inline skyportal/handlers/api/gwdetector.py:50]: Isn't this taken care of by the schema validation?

File: skyportal/handlers/api/gwdetector.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        multiple:
+          description: Retrieve all gwdetectors
+          tags:
+            - gwdetectors
+          parameters:
+            - in: query
+              name: name
+              schema:
+                type: string
+              description: Filter by name (exact match)
```
Comment (guynir42) [inline skyportal/handlers/api/gwdetector.py:104]: A partial match could be a real time saver in this case.

File: skyportal/handlers/api/gwdetector.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+        with self.Session() as session:
+            t = session.scalars(
+                GWDetector.select(session.user_or_token, mode="update").where(
+                    GWDetector.id == int(gwdetector_id)
+                )
+            ).first()
+            if t is None:
+                return self.error('Invalid GW Detector ID.')
+            data = self.get_json()
```
Comment (guynir42) [inline skyportal/handlers/api/gwdetector.py:174]: A PUT should let you modify or create if doesn't exist. Maybe you want to PATCH in this case?

File: skyportal/handlers/api/gwdetector.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+                )
+
+            if 'name' in data:
+                t.name = data['name']
+            if 'nickname' in data:
+                t.nickname = data['nickname']
+            if 'lat' in data:
+                t.lat = data['lat']
+            if 'lon' in data:
+                t.lon = data['lon']
```
Comment (guynir42) [inline skyportal/handlers/api/gwdetector.py:192]: Should definitely test that lon/lat are in range.

File: skyportal/models/gwdetector.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+def manage_gwdetector_access_logic(cls, user_or_token):
+    if user_or_token.is_system_admin:
+        return DBSession().query(cls)
+    elif 'Manage allocations' in [acl.id for acl in user_or_token.acls]:
+        return DBSession().query(cls)
+    else:
+        # return an empty query
+        return DBSession().query(cls).filter(cls.id == -1)
+
```
Comment (guynir42) [inline skyportal/models/gwdetector.py:25]: We are going to need to figure out how to do this without DBSession()

File: skyportal/models/gwdetector.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        sa.String,
+        unique=True,
+        nullable=False,
+        doc="Unabbreviated facility name (e.g., LIGO Hanford Observatory.",
+    )
+    nickname = sa.Column(
+        sa.String, nullable=False, doc="Abbreviated facility name (e.g., LHO)."
+    )
+    lat = sa.Column(sa.Float, nullable=False, doc='Latitude in deg.')
+    lon = sa.Column(sa.Float, nullable=False, doc='Longitude in deg.')
```
Comment (guynir42) [inline skyportal/models/gwdetector.py:43]: Not linking the GCNs with this in some relationship?

File: static/js/components/GWDetectorPageDesktop.jsx [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    ...(gwdetector?.elevation ? [`Elevation: ${gwdetector.elevation}`] : []),
+  ];
+
+  // eslint-disable-next-line prefer-template
+  const result = "( " + array.join(" / ") + " )";
+
+  return result;
+}
+
+const GWDetectorPage = () => {
```
Comment (guynir42) [inline static/js/components/GWDetectorPageDesktop.jsx:76]: Isn't there already a component called `GWDetectorPage`?

File: static/js/ducks/gwdetector.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        ...state,
+        ...gwdetector,
+      };
+    }
+    case CURRENT_GWDETECTORS_AND_MENU: {
+      const gwdetector = {};
+      console.log(
+        "action.data.currentGWDetectorMenu",
+        action.data.currentGWDetectorMenu
+      );
```
Comment (guynir42) [inline static/js/ducks/gwdetector.js:53]:
```suggestion
```

File: static/js/ducks/gwdetectors.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        ...state,
+        gwdetectorList,
+      };
+    }
+    default:
+      return state;
+  }
+};
+
+store.injectReducer("gwdetectors", reducer);
```
Comment (guynir42) [inline static/js/ducks/gwdetectors.js:36]: I would much rather see a `FETCH_GWDETECTORS_LIST` and so on, in the same file with the singular reducers. 
Having gwdetector and gwdetectors is not good design.

**Review by guynir42 — COMMENTED** [review]
I have some comments. Most of them pretty minor.

File: skyportal/handlers/api/gwdetector.py (hunk 1 above)
Comment (mcoughlin) [reply skyportal/handlers/api/gwdetector.py:9]: At this point, allocations is kind of a catch all for instrument related stuff. What do you think?

File: skyportal/handlers/api/gwdetector.py (hunk 2 above)
Comment (mcoughlin) [reply skyportal/handlers/api/gwdetector.py:50]: Well it catches it pre-schema. You want it gone?

File: skyportal/handlers/api/gwdetector.py (hunk 3 above)
Comment (mcoughlin) [reply skyportal/handlers/api/gwdetector.py:104]: good idea.

File: skyportal/handlers/api/gwdetector.py (hunk 4 above)
Comment (mcoughlin) [reply skyportal/handlers/api/gwdetector.py:174]: good point.

File: skyportal/handlers/api/gwdetector.py (hunk 5 above)
Comment (mcoughlin) [reply skyportal/handlers/api/gwdetector.py:192]: good idea.

File: skyportal/models/gwdetector.py (hunk 6 above)
Comment (mcoughlin) [reply skyportal/models/gwdetector.py:25]: yeah this is borrowed from the other handlers. We can fix this in a future PR all at once.

File: skyportal/models/gwdetector.py (hunk 7 above)
Comment (mcoughlin) [reply skyportal/models/gwdetector.py:43]: Not yet, that should be in the next few PRs.

File: static/js/components/GWDetectorPageDesktop.jsx (hunk 8 above)
Comment (mcoughlin) [reply static/js/components/GWDetectorPageDesktop.jsx:76]: good point.

File: static/js/ducks/gwdetector.js (hunk 9 above)
Comment (mcoughlin) [reply static/js/ducks/gwdetector.js:53]: oops.

File: static/js/ducks/gwdetectors.js (hunk 10 above)
Comment (mcoughlin) [reply static/js/ducks/gwdetectors.js]: good point

File: skyportal/handlers/api/gwdetector.py (hunk 1 above)
Comment (guynir42) [reply skyportal/handlers/api/gwdetector.py:9]: Not critical. Should consider if we want to add another permission to cover these things.

File: skyportal/handlers/api/gwdetector.py (hunk 2 above)
Comment (guynir42) [reply skyportal/handlers/api/gwdetector.py:50]: If it is indeed redundant I suggest we remove it. Will save you some time in the PUT/PATCH handler.

File: skyportal/models/gwdetector.py (hunk 6 above)
Comment (guynir42) [reply skyportal/models/gwdetector.py:25]: Agreed.

File: skyportal/handlers/api/gwdetector.py (hunk 1 above)
Comment (mcoughlin) [reply skyportal/handlers/api/gwdetector.py:9]: My vote would be to leave this permission in but basically rename it.

Comment (mcoughlin) [reply skyportal/handlers/api/gwdetector.py:9]: (in a future PR)

File: skyportal/handlers/api/gwdetector.py (hunk 2 above)
Comment (mcoughlin) [reply skyportal/handlers/api/gwdetector.py]: done.

File: skyportal/models/mmadetector.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    create = update = delete = CustomUserAccessControl(manage_mmadetector_access_logic)
+
+    name = sa.Column(
+        sa.String,
+        unique=True,
+        nullable=False,
+        doc="Unabbreviated facility name (e.g., LIGO Hanford Observatory.",
+    )
+    nickname = sa.Column(
+        sa.String, nullable=False, doc="Abbreviated facility name (e.g., LHO)."
```
Comment (guynir42) [inline skyportal/models/mmadetector.py:48]: E.g., should be H1 maybe?

File: static/js/ducks/mmadetector.js [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+const REFRESH_MMADETECTOR = "skyportal/REFRESH_MMADETECTOR";
+
+const FETCH_MMADETECTOR = "skyportal/FETCH_MMADETECTOR";
+const FETCH_MMADETECTOR_OK = "skyportal/FETCH_MMADETECTOR_OK";
+
+const SUBMIT_MMADETECTOR = "skyportal/SUBMIT_MMADETECTOR";
+
+const CURRENT_MMADETECTORS_AND_MENU = "skyportal/CURRENT_MMADETECTORS_AND_MENU";
+
+const REFRESH_MMADETECTOR_LIST = "skyportal/REFRESH_MMADETECTORS";
```
Comment (guynir42) [inline static/js/ducks/mmadetector.js:15]:
```suggestion
const REFRESH_MMADETECTOR_LIST = "skyportal/REFRESH_MMADETECTOR_LIST";
```

**Review by guynir42 — APPROVED** [review]
Minor comments. Looks good.

File: skyportal/handlers/api/gcn.py [hunk 13]
```diff
@@ -115,9 +115,17 @@ def post_gcnevent_from_xml(payload, user_id, session):
         sent_by_id=user.id,
     )
 
+    detectors = []
     for tag in tags:
         session.add(tag)
+
+        mma_detector = session.scalars(
+            MMADetector.select(user).where(MMADetector.nickname == tag.text)
+        ).first()
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/gcn.py:124]: Yeah we will probably evolve this when the final O4 alert schema gets set. But it's also a common GW thing, so idk.

File: skyportal/models/mmadetector.py (hunk 11 above)
Comment (mcoughlin) [reply skyportal/models/mmadetector.py:48]: good point.

**Pull request merged**